### DISCUSSION TITLE
Suggestion about Themes  playground

### DISCUSSION THREAD

**Discussion by zakariachamakh** [rydmike/flex_color_scheme#45] [Ideas]
# Sketch  of the idea :  

![flex_color_schema](https://user-images.githubusercontent.com/53184374/151543740-4e6fe30e-afac-484c-9b26-8608d7baaea5.jpg)

- I have checked your theme by [playground](https://rydmike.com/flexcolorschemeV4Tut5/#/) and I think we can assess the user experience and  make more easy if we add the feature that  i explained in  the sketch above .
 
- if you think its a useful idea please contact meso we can collaborate .

**Comment by rydmike** [comment]
Hi @zakariachamakh, thanks for this suggestion and for also using the Themes Playground part of FlexColorScheme and your fabulous feedback on it.

I totally agree with you that for a nicer user experience in a more sophisticated theming helper app, this would provide a more guided and better workflow for users of the app that are not familiar with all the properties.

The app was actually never intended to be a theming app the way it is now, it is a total bonus side-effect, hehe. It was and still is mainly just the last step in the readme tutorial (example 5), used to show what kind of themes you can make and build with the FlexColorScheme package, plus some other simple techniques that it also demonstrates by using the official Flutter skeleton template architecture.

The main purpose is the FlexColorScheme package itself, not, example 5, that kind of evolved into the Themes Playground. It is just a small sidekick, admittedly one that became much more fancy than I ever originally intended 😄 

I feel I should not complicate the example 5 app a lot further, at least not if users are intended to be able to follow what is going on it. I also have plans to write longer tutorials using the examples, when I get the readme moved over to a separate documentation site. The current readme has reached the max length allowed on pub.dev (128kB). If I complicate the example app 5 a lot further, then writing the more complete tutorial for becomes a very big task. 

If the focus would have been to build a very user friendly theme configurator companion app to FlexColorScheme, or to add  one that is. I would definitely do something very similar to what you suggest, I do like it a lot. I would probably also add compact view that shows all widgets with their themes applied, as step after configuring them all, before the step to get the theme code.

The sliding built in color selector could probably remain on top of the config screen, so you can always change the colors, and have the panels that show all the color details as one of the 'selectable' widgets that you can go back to, it would just start with it selected.

For my own/advanced use/preference, I would probably keep the current style around as an "advanced config" optional view in the app. The reason why I actually like this current style (apart from it being a bit simpler to build as an example app), is because on a large 32" 4k screen, I can see and configure everything in one big view without moving back and forth between a lot of widget based views. I can also directly change any config anywhere, and observe the result on all widgets all at once after any change. Makes it a lot faster for me and I see all the results at once all the time. However, you need a 4k screen running at native resolution for that, and it needs imo to be at least 32" size too, because 4k gets to small unscaled otherwise. So sure, not everybody has that kind of desktop, but I like it when I configure my own themes, hehe 😃 

I do agree with you though, that for most users, with typical setups, and for a better and more guided theming flow, and for an app really designed as a theming companion app to FlexColorScheme (which this app was not originally intended to be), it would be a really nice design and welcome addition to make something along the lines you propose.

One way to perhaps try it, would be to make a branch where I setup a copy of example 5 as e.g. "theme_configurator" (or something similarly named) and we try out some ideas together. 

If making a broader app like this, I might also consider adding Riverpod, to make state management a bit cleaner. But OK, even using current approach and passing the theme controller around might still be ok, although we might want to handle a bit more things in the app state for the more complex UI, that does not fit in the theme controller so adding Riverpod would help then, just minor details.

Let me think about it a bit more and get back to you. Meanwhile feel free to add more of your thoughts and ideas on the topic. 😃

**Comment by zakariachamakh** [comment]
- Hi  @rydmike  first of all thank you for replying and am really glad that you had a look on my sketch and you liked it well I understand that flexcolor is the main subject it's really a great idea to work on but in my opinion we can push it further and make it easier and faster by making the theming app top of it .am really excited to collaborate with you if you want to start the theming app because am always facing this repeating problem while implementing the design of any project and many of us suffer from the same issue

- So to get straight to the point my idea is providing a graphical ui to customize the material design or any design system where we will custom the fontography , colors, shadows and make it easier and faster

- We can take inspiration from  [**sprint ui**](https://sprintui.com/)  also [**components**](https://components.ai/box-shadow/h71Mh0o3V0e4ecRO20Vu) how they are presenting their artibut of components . We can also have a look on [**eva design**](https://colors.eva.design/)  for the first step of our user flow and how the site is presenting the color palette . Also we can benefit from [**flutter flow**](https://flutterflow.io/) not in the approach of coding but in a way that will serve us

- You also mentioned the  point  of the state management (riverpod) but I personally fan of [**getx**](https://pub.dev/packages/get) I think it's more  scalable .

**Comment by rydmike** [comment]
Wow, that is both cool and very ambitious. It goes quite far beyond the scope of FlexColorScheme and also beyond what I can commit time to I'm afraid. I think it is definitely something that should be built outside the FlexColorScheme package repo as its own solution.

If you have the time and passion to build such a tool for Flutter, I recommend just start doing it. I think it would be very welcome and appreciated by the Flutter community and definitely a great project to use to learn a lot about Flutter with.

When it comes to packages and tools you prefer to use in such a project, I do recommend a bit more research. Personally I don't use GetX for reasons that are quite well summarized here https://twitter.com/scottstoll2017/status/1468431032240246784

Still if you like it, then sure it works and the nice part with personal projects, is just that, you can use the tools you like and prefer 😃 Be aware that GetX is perhaps not so well regarded by mainstream Flutter community. That said, it certainly works and is convenient in many ways, it is just not very Flutter like in its nature. Plus many developers do not like the it bundles so many features and functions in one package, they feel you put to many eggs in one basket when you use it. It is like its own little microframework on top of Flutter. Then again I have also seen that some devs likes this convenience, so it is definitely a mixed bag.

> **SIDENOTE**
>You might already know it, but I also do this chart on Twitter that lists major state management packages: https://twitter.com/RydMike/status/1487902422672580608 I will probably publish a new version soon.

**Comment by rydmike** [comment]
This is coming in version 5.0.0, it will be available very soon a v5.0.0-dev.2 before stable release:

![image](https://user-images.githubusercontent.com/39990307/161347145-21a556dd-d12f-499f-b34c-3945ecc3e83b.png)
![image](https://user-images.githubusercontent.com/39990307/161347160-455b7aec-a9ec-4776-b6bf-70a2a62b1caa.png)
![image](https://user-images.githubusercontent.com/39990307/161347302-d0ab91ea-c5b0-4db3-81dd-a8d45cf644ec.png)

It is not exactly as you wished, but similar and made with smaller adjustments from the grid view to a page view, works down to phone size too

![image](https://user-images.githubusercontent.com/39990307/161347460-86a0ba79-79a4-48f8-a507-4cada9c68ca4.png)

The large grid view still exists as well, one can switch between them:

![image](https://user-images.githubusercontent.com/39990307/161347543-d54e4408-7ce9-4352-9d32-55dfde62a200.png)

----

You will find a live web version of it soon and the new dev version on pub.dev. I'll drop you not again then.

**Comment by rydmike** [comment]
Hi again @zakariachamakh, you can now find this version of the Themes Playground app here: https://rydmike.com/flexcolorscheme/themesplayground-v5

It is the companion for the new **FlexColorScheme version 5**, which is still only out as pre-release here: https://pub.dev/packages/flex_color_scheme/versions/5.0.0-dev.2

It will be released as stable when the documentation update has been completed.
